rotkehlchen/chain/evm/decoding/uniswap/utils.py

```diff
@@ -128,3 +141,97 @@ def get_uniswap_swap_amounts(tx_log: 'EvmTxReceiptLog') -> tuple[int, int]:
+    sqrt = LOG_PRICE**FVal(tick / 2) * POW_96
+    sqrt = max(min(sqrt, sqrt_b), sqrt_a)
+
+    return sqrt, sqrt_a, sqrt_b
+
+
+def calculate_amount(
+        tick_lower: int,
+        liquidity: int,
+        tick_upper: int,
+        decimals: int,
+        tick: int,
+        token_position: Literal[0, 1],
+) -> FVal:
+    """
+    Calculates the amount of a token in the Uniswap V3 LP position.
+    `token_position` can either be 0 or 1 representing the position of the token in a pair.
+    """
+    sqrt, sqrt_a, sqrt_b = compute_sqrt_values_for_amounts(
+        tick_lower=tick_lower,
+        tick_upper=tick_upper,
+        tick=tick,
+    )
+    if token_position == 0:
+        amount = (liquidity * POW_96 * (sqrt_b - sqrt) / (sqrt_b * sqrt)) / 10**decimals
+    elif token_position == 1:
+        amount = liquidity * (sqrt - sqrt_a) / POW_96 / 10**decimals
+
+    return FVal(amount)
+
+
+def get_position_price_from_underlying(
+        evm_inquirer: 'EvmNodeInquirer',
+        token0_raw_address: str,
+        token1_raw_address: str,
+        tick_lower: int,
+        tick_upper: int,
+        liquidity: int,
+        tick: int,
+        price_func: Callable[['Asset'], Price],
+) -> Price:
+    """Get a Uniswap LP position's price from its underlying assets and tick/liquidity info."""
+    position_price = ZERO
+    for idx, raw_token_address in enumerate((token0_raw_address, token1_raw_address)):
+        try:
+            token_address = deserialize_evm_address(raw_token_address)
+        except DeserializationError as e:
+            log.error(
+                f'Encountered invalid address {raw_token_address} while '
+                f'getting Uniswap LP position price on {evm_inquirer.chain_name}. {e!s}',
+            )
+            return ZERO_PRICE
+
+        if token_address == ZERO_ADDRESS:
+            underlying_asset = evm_inquirer.native_token
+        else:
+            try:
+                underlying_asset = get_or_create_evm_token(
+                    userdb=evm_inquirer.database,
+                    evm_inquirer=evm_inquirer,
+                    evm_address=deserialize_evm_address(token_address),
+                    chain_id=evm_inquirer.chain_id,
+                )
+            except (NotERC20Conformant, NotERC721Conformant, InputError) as e:
+                log.error(
+                    f'Failed to fetch {evm_inquirer.chain_name} token {token_address} while '
+                    f'getting Uniswap LP position price due to: {e!s}',
+                )
+                return ZERO_PRICE
+
+        asset_usd_price = price_func(underlying_asset)
+        if asset_usd_price == ZERO_PRICE:
+            continue
+
+        position_price += FVal(calculate_amount(
+            tick_lower=tick_lower,
+            liquidity=liquidity,
+            tick_upper=tick_upper,
+            decimals=get_decimals(underlying_asset),
+            tick=tick,
+            token_position=idx,  # type: ignore  # idx will only be 0 or 1
```

Comment:
mmmm I understand that the way you wrote the code is the one that best captures the concept but I'm seeing that you repeat the calcs here. 

you could calculate 

```
    sqrt, sqrt_a, sqrt_b = compute_sqrt_values_for_amounts(
        tick_lower=tick_lower,
        tick_upper=tick_upper,
        tick=tick,
    )
```

in this function and then pass them as arguments to `calculate_amount` to obtain the two values. Notice that if get_decimals is the same value for both tokens you are just asking for both branches in `calculate_amount`. Furthermore `decimals` is only needed for the position 1 right? so you don't need to mess retrieving the decimals for the other token at all. Feels like this whole function and call chain can be optimized a lot to do much less work